bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartWebSocket.java

```diff
@@ -121,11 +122,12 @@ public void onConnect(Session session) {
 
         ScheduledFuture<?> connectionTracker = this.connectionTracker;
         if (connectionTracker != null && !connectionTracker.isCancelled()) {
-            connectionTracker.cancel(false);
+            connectionTracker.cancel(true);
+            this.connectionTracker = null;
```

Comment:
Because of line 130 this line (126) is not needed.

Reply:
Something is fishy. Nullifying and reassigning should make no difference for the Garbage Collector, in both cases the object is eligible for GC.

I'm wondering if actually the bug could be something else? For example, previously `cancel()` was called without interrupting the thread, so it could still be running, but marked as cancelled. On next call another thread could be spawned, while old one is still running (but without any references to it). This could happen a few times depending on how long these threads would continue running. At some point, could this lead to a problem, or is it too theoretical? Just adding some lose thoughts, hopefully they will trigger new ideas and not mislead too much. :)

Reply:
@jlaur I share your concerns; I see that @gerrieg also changed `cancel(false)` to `cancel(true)` so that may be his actual “fix”; ?? .. but I think it is important that we understand what is really going on here; let me think about this further; I will get back tomorrow..

Reply:
> explicitly setting it to null helps the garbage collector and the scheduled pool

I am pretty sure the above hypothesis is wrong: the binding code only nulls an external reference to the task; the task has already been submitted to the OH scheduler, and that holds its own internal reference until the task is finished; so nulling this external reference cannot reduce the reference count sufficiently to trigger the GC.

> the connection tracker will not start after some restarts.

So we need to get to the root of why this happens. (And more specifically why you see it on on your system, but I do not see it on my own system).

Maybe the scheduled task method `sendKeepAlivePing()` is blocking in some way; perhaps on a TCP/IP IO call. In that case, the change from `cancel(false)` to `cancel(true)` might already be sufficient to fix things (because it would interrupt the IO call on that respective thread). 

It also occurs to me that the `sendKeepAlivePing()` could potentially have been called on two tasks from two separate scheduled task series -- the first one having blocked in the first task series, and the second one therefore not being thread safe. So to prevent that we should (in any case) synchronize the scheduled task method i.e. `private synchronized void sendKeepAlivePing()`.

Note: if there were some IO blocking going on, then synchronize would actually make the error more obvious, but at least it would a first step towards identifing what might be going on underneath..